File: src/ios/Sources/GeolocationPlugin.swift
import CoreLocation
import Foundation

/**
 * The device's location, as a request you start and a result you collect.
 *
 * Nothing here blocks. `CLLocationManager` delivers to its delegate on the main
 * queue, so a call that waited for a fix on the calling thread would wait for
 * something that can never arrive there — the Rust side starts a request and
 * polls for the answer instead.
 *
 * The permission itself is declared by the Dioxus CLI from `[permissions]` in
 * `Dioxus.toml`, which writes `NSLocationWhenInUseUsageDescription`. Without
 * that string in the Info.plist iOS refuses the request outright rather than
 * showing a prompt.
 */
@objc(GeolocationPlugin)
public class GeolocationPlugin: NSObject, CLLocationManagerDelegate {

    private static let stateIdle = "idle"
    private static let stateLocating = "locating"

    private let lock = NSLock()
    private var locating = false
    private var result: String?

    /// Read from any thread, written only on the main one. `CLLocationManager`
    /// is main-thread state, and a synchronous hop from the Rust thread to read
    /// it could deadlock against a main thread waiting on that same call, so
    /// the status is mirrored here instead. The first poll after launch can see
    /// the pre-launch default; the second sees the truth.
    private var cachedStatus: CLAuthorizationStatus = .notDetermined

    private var manager: CLLocationManager?
    private var timeoutWork: DispatchWorkItem?
    private var pendingHighAccuracy = false

    override public init() {
        super.init()
        // Get the manager and its authorization status on screen early, so the
        // first permission poll from Rust already has a real answer.
        DispatchQueue.main.async { self.ensureManager() }
    }

    // MARK: - Permissions

    @objc
    public func checkPermissionsFromRust() -> String {
        let state: String
        switch cachedStatus {
        case .notDetermined: state = "prompt"
        case .restricted, .denied: state = "denied"
        case .authorizedAlways, .authorizedWhenInUse: state = "granted"
        @unknown default: state = "prompt"
        }
        // iOS grants location as one thing; it has no separate coarse grant the
        // way Android does, so both aliases report the same state.
        let payload = ["location": state, "coarseLocation": state]
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8) else {
            return "{\"location\":\"prompt\",\"coarseLocation\":\"prompt\"}"
        }
        return json
    }

    @objc
    public func requestPermissionsFromRust() {
        DispatchQueue.main.async {
            guard let manager = self.ensureManager() else { return }
            // iOS shows this once per install. After a refusal it does nothing
            // at all, which is why the Rust side treats denied as final.
            manager.requestWhenInUseAuthorization()
        }
    }

    // MARK: - Position

    @objc
    public func startPositionRequestFromRust(_ optionsJson: String) {
        var highAccuracy = false
        var timeoutMs: Double = 10_000
        var maximumAgeMs: Double = 0
        if let data = optionsJson.data(using: .utf8),
           let options = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] {
            highAccuracy = options["enableHighAccuracy"] as? Bool ?? false
            timeoutMs = options["timeout"] as? Double ?? 10_000
            maximumAgeMs = options["maximumAge"] as? Double ?? 0
        }

        DispatchQueue.main.async {
            self.cancelInFlight()
            self.lock.lock()
            self.result = nil
            self.locating = true
            self.lock.unlock()

            guard let manager = self.ensureManager() else {
                self.finish(Self.errorJson("Location services are unavailable."))
                return
            }

            if maximumAgeMs > 0, let cached = manager.location,
               Date().timeIntervalSince(cached.timestamp) * 1000 <= maximumAgeMs {
                self.finish(Self.positionJson(cached))
                return
            }

            manager.desiredAccuracy = highAccuracy
                ? kCLLocationAccuracyBest
                : kCLLocationAccuracyHundredMeters

            let expire = DispatchWorkItem { [weak self] in
                self?.finish(Self.errorJson("Timed out waiting for a location."))
            }
            self.timeoutWork = expire
            DispatchQueue.main.asyncAfter(
                deadline: .now() + max(timeoutMs / 1000, 0.1),
                execute: expire
            )

            if self.cachedStatus == .notDetermined {
                // Asking now and requesting on the authorization callback,
                // rather than firing a request iOS would silently drop.
                self.pendingHighAccuracy = true
                manager.requestWhenInUseAuthorization()
                return
            }
            guard self.cachedStatus == .authorizedAlways
                || self.cachedStatus == .authorizedWhenInUse else {
                self.finish(Self.errorJson("Location permission has not been granted."))
                return
            }
            manager.requestLocation()
        }
    }

    @objc
    public func getLocationStateFromRust() -> String {
        lock.lock()
        defer { lock.unlock() }
        return locating ? Self.stateLocating : Self.stateIdle
    }

    @objc
    public func takePositionFromRust() -> String? {
        lock.lock()
        defer { lock.unlock() }
        let taken = result
        result = nil
        return taken
    }

    // MARK: - Delegate

    public func locationManager(
        _ manager: CLLocationManager,
        didUpdateLocations locations: [CLLocation]
    ) {
        guard let location = locations.last else { return }
        finish(Self.positionJson(location))
    }

    public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        finish(Self.errorJson("Could not get a location: \(error.localizedDescription)"))
    }

    public func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        cachedStatus = manager.authorizationStatus
        guard pendingHighAccuracy else { return }
        switch cachedStatus {
        case .notDetermined:
            // Still on screen; wait for the user to answer.
            return
        case .authorizedAlways, .authorizedWhenInUse:
            pendingHighAccuracy = false
            manager.requestLocation()
        default:
            pendingHighAccuracy = false
            finish(Self.errorJson("Location permission was refused."))
        }
    }

    // MARK: - Internals, all on the main thread

    @discardableResult
    private func ensureManager() -> CLLocationManager? {
        if let existing = manager {
            cachedStatus = existing.authorizationStatus
            return existing
        }
        let created = CLLocationManager()
        created.delegate = self
        manager = created
        cachedStatus = created.authorizationStatus
        return created
    }

    /// Settle the request, whichever way it went, and stop listening.
    private func finish(_ json: String) {
        cancelInFlight()
        lock.lock()
        result = json
        locating = false
        lock.unlock()
    }

    private func cancelInFlight() {
        timeoutWork?.cancel()
        timeoutWork = nil
        pendingHighAccuracy = false
    }

    private static func errorJson(_ message: String) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: ["error": message]),
              let json = String(data: data, encoding: .utf8) else {
            return "{\"error\":\"Could not get a location.\"}"
        }
        return json
    }

    private static func positionJson(_ location: CLLocation) -> String {
        var coords: [String: Any] = [
            "latitude": location.coordinate.latitude,
            "longitude": location.coordinate.longitude,
            // A negative accuracy means the reading is invalid rather than
            // very precise, so report it as unknown rather than as a number.
            "accuracy": max(location.horizontalAccuracy, 0),
        ]
        if location.verticalAccuracy >= 0 {
            coords["altitude"] = location.altitude
            coords["altitudeAccuracy"] = location.verticalAccuracy
        }
        if location.speed >= 0 {
            coords["speed"] = location.speed
        }
        if location.course >= 0 {
            coords["heading"] = location.course
        }
        let payload: [String: Any] = [
            "timestamp": Int(location.timestamp.timeIntervalSince1970 * 1000),
            "coords": coords,
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8) else {
            return errorJson("Could not encode the location.")
        }
        return json
    }
}

File: src/android/geolocation/src/main/kotlin/dev/dioxus/g3_native_plugins/geolocation/GeolocationPlugin.kt
package dev.dioxus.g3_native_plugins.geolocation

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import org.json.JSONObject

/**
 * The device's location, as a request you start and a result you collect.
 *
 * A fix takes seconds and arrives on a callback, so nothing here blocks: the
 * Rust side starts a request and polls for the answer. Blocking the calling
 * thread would stall a Dioxus effect for as long as the fix took, and the
 * listener needs a live Looper to be delivered on anyway.
 *
 * Fixes come from the platform [LocationManager] rather than Google's fused
 * provider. The fused provider gives better results for less battery, but only
 * at the cost of forcing play-services-location on everyone who depends on this
 * crate, which is not a trade a library gets to make for its consumers.
 *
 * Requests listen to every enabled provider at once and take the first fix.
 * See [providers] for why picking only the best-looking one is a trap.
 *
 * The permission itself is declared by the Dioxus CLI from `[permissions]` in
 * `Dioxus.toml`. This asks the user for it at runtime, which Android requires
 * separately from declaring it.
 */
class GeolocationPlugin(private val activity: Activity) {
    companion object {
        private const val REQUEST_CODE = 7301
        private const val STATE_IDLE = "idle"
        private const val STATE_LOCATING = "locating"
    }

    private val manager: LocationManager? =
        activity.getSystemService(Context.LOCATION_SERVICE) as? LocationManager

    private val lock = Object()
    private var locating = false
    private var result: String? = null
    private var listener: LocationListener? = null
    private var timeout: Runnable? = null
    private val handler = Handler(Looper.getMainLooper())

    // ---- Permissions ----

    private fun stateOf(permission: String): String = when {
        activity.checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED -> "granted"
        // Android says to explain first only once the user has already refused,
        // so this doubles as "asked and declined, but not permanently".
        activity.shouldShowRequestPermissionRationale(permission) -> "prompt-with-rationale"
        else -> "prompt"
    }

    fun checkPermissionsFromRust(): String = JSONObject()
        .put("location", stateOf(Manifest.permission.ACCESS_FINE_LOCATION))
        .put("coarseLocation", stateOf(Manifest.permission.ACCESS_COARSE_LOCATION))
        .toString()

    fun requestPermissionsFromRust() {
        activity.runOnUiThread {
            val wanted = arrayOf(
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION,
            ).filter {
                activity.checkSelfPermission(it) != PackageManager.PERMISSION_GRANTED
            }
            if (wanted.isEmpty()) return@runOnUiThread
            // The result arrives on onRequestPermissionsResult, an Activity
            // callback a library cannot override from outside. The Rust side
            // polls checkPermissions instead, which is cheap and needs no hook.
            activity.requestPermissions(wanted.toTypedArray(), REQUEST_CODE)
        }
    }

    private fun hasAnyLocationPermission(): Boolean =
        activity.checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) ==
            PackageManager.PERMISSION_GRANTED ||
            activity.checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) ==
            PackageManager.PERMISSION_GRANTED

    // ---- Position ----

    fun startPositionRequestFromRust(optionsJson: String) {
        val options = try {
            JSONObject(optionsJson)
        } catch (error: Exception) {
            JSONObject()
        }
        val highAccuracy = options.optBoolean("enableHighAccuracy", false)
        val timeoutMs = options.optLong("timeout", 10_000L).coerceAtLeast(1L)
        val maximumAgeMs = options.optLong("maximumAge", 0L)

        activity.runOnUiThread {
            // Registration, cancellation and delivery all belong on a thread
            // with a Looper; Rust effects run on a native worker thread.
            cancelInFlight()
            synchronized(lock) {
                result = null
                locating = true
            }

            if (!hasAnyLocationPermission()) {
                finish(errorJson("Location permission has not been granted."))
                return@runOnUiThread
            }
            val manager = manager
            if (manager == null) {
                finish(errorJson("Location services are unavailable."))
                return@runOnUiThread
            }

            cachedLocation(manager, maximumAgeMs)?.let {
                finish(positionJson(it))
                return@runOnUiThread
            }

            val providers = providers(manager, highAccuracy)
            if (providers.isEmpty()) {
                finish(errorJson("No location provider is enabled."))
                return@runOnUiThread
            }

            val created = LocationListener { location ->
                // First fix wins, whichever provider produced it. finish()
                // removes this listener from all of them.
                finish(positionJson(location))
            }
            listener = created
            val expire = Runnable { finish(errorJson("Timed out waiting for a location.")) }
            timeout = expire
            handler.postDelayed(expire, timeoutMs)

            var listening = 0
            var lastFailure: String? = null
            for (provider in providers) {
                try {
                    @Suppress("MissingPermission")
                    manager.requestLocationUpdates(
                        provider,
                        0L,
                        0f,
                        created,
                        Looper.getMainLooper(),
                    )
                    listening += 1
                } catch (error: SecurityException) {
                    lastFailure = "Location permission was revoked: ${error.message}"
                } catch (error: Exception) {
                    lastFailure = "Could not start location updates: ${error.message}"
                }
            }
            if (listening == 0) {
                finish(errorJson(lastFailure ?: "Could not start location updates."))
            }
        }
    }

    fun getLocationStateFromRust(): String =
        synchronized(lock) { if (locating) STATE_LOCATING else STATE_IDLE }

    fun takePositionFromRust(): String? = synchronized(lock) {
        val taken = result
        result = null
        taken
    }

    // ---- Internals, all on the main thread ----

    /** Settle the request, whichever way it went, and stop listening. */
    private fun finish(json: String) {
        cancelInFlight()
        synchronized(lock) {
            result = json
            locating = false
        }
    }

    private fun cancelInFlight() {
        listener?.let { current ->
            try {
                manager?.removeUpdates(current)
            } catch (_: Exception) {
                // Removing an already-removed listener is not worth reporting.
            }
        }
        listener = null
        timeout?.let(handler::removeCallbacks)
        timeout = null
    }

    /**
     * Every provider worth listening to, rather than the single best one.
     *
     * A provider can report itself enabled and still never produce a fix:
     * network location with nothing to work from, or an emulator where only GPS
     * is fed. Betting a whole request on one provider turns that into a timeout
     * while another provider had an answer sitting ready — which is exactly what
     * happened the first time this ran on a device. Registering on all of them
     * and taking whichever answers first costs nothing and cannot do worse.
     *
     * High accuracy is the exception. Asking for GPS-grade precision and then
     * returning a coarse network fix because it arrived first would not be the
     * thing that was asked for, so that case listens to GPS alone.
     */
    private fun providers(manager: LocationManager, highAccuracy: Boolean): List<String> {
        val enabled = listOf(LocationManager.GPS_PROVIDER, LocationManager.NETWORK_PROVIDER)
            .filter { name ->
                try {
                    manager.isProviderEnabled(name)
                } catch (_: Exception) {
                    false
                }
            }
        if (highAccuracy && enabled.contains(LocationManager.GPS_PROVIDER)) {
            return listOf(LocationManager.GPS_PROVIDER)
        }
        return enabled
    }

    private fun cachedLocation(manager: LocationManager, maximumAgeMs: Long): Location? {
        if (maximumAgeMs <= 0L) return null
        val oldestAcceptable = System.currentTimeMillis() - maximumAgeMs
        return manager.allProviders
            .mapNotNull { name ->
                try {
                    @Suppress("MissingPermission")
                    manager.getLastKnownLocation(name)
                } catch (_: Exception) {
                    null
                }
            }
            .filter { it.time >= oldestAcceptable }
            .maxByOrNull { it.time }
    }

    private fun errorJson(message: String): String =
        JSONObject().put("error", message).toString()

    private fun positionJson(location: Location): String {
        val coords = JSONObject()
            .put("latitude", location.latitude)
            .put("longitude", location.longitude)
            .put("accuracy", location.accuracy.toDouble())
        if (location.hasAltitude()) coords.put("altitude", location.altitude)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O &&
            location.hasVerticalAccuracy()
        ) {
            coords.put("altitudeAccuracy", location.verticalAccuracyMeters.toDouble())
        }
        if (location.hasSpeed()) coords.put("speed", location.speed.toDouble())
        if (location.hasBearing()) coords.put("heading", location.bearing.toDouble())
        return JSONObject()
            .put("timestamp", location.time)
            .put("coords", coords)
            .toString()
    }
}
